pkg/collector/corechecks/snmp/snmp.go

```diff
@@ -0,0 +1,141 @@
+package snmp
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"time"
+)
+
+const (
+	snmpCheckName = "snmp"
+)
+
+// Check aggregates metrics from one Check instance
+type Check struct {
+	core.CheckBase
+	config  snmpConfig
+	session sessionAPI
+	sender  metricSender
+}
+
+// Run executes the check
+func (c *Check) Run() error {
+	start := time.Now()
+
+	sender, err := aggregator.GetSender(c.ID())
+	if err != nil {
+		return err
+	}
+	c.sender = metricSender{sender: sender}
+
+	staticTags := c.config.getStaticTags()
+
+	var checkErr error
+	tags, checkErr := c.doRun(staticTags)
+	if checkErr != nil {
+		c.sender.serviceCheck("snmp.can_check", metrics.ServiceCheckCritical, "", tags, checkErr.Error())
+	} else {
+		c.sender.serviceCheck("snmp.can_check", metrics.ServiceCheckOK, "", tags, "")
+	}
+
+	c.sender.gauge("snmp.devices_monitored", float64(1), "", tags)
+
+	// SNMP Performance metrics
+	c.sender.monotonicCount("datadog.snmp.check_interval", float64(start.UnixNano())/1e9, "", tags)
+	c.sender.gauge("datadog.snmp.check_duration", float64(time.Since(start))/1e9, "", tags)
+	c.sender.gauge("datadog.snmp.submitted_metrics", float64(c.sender.submittedMetrics), "", tags)
+
+	// Commit
+	sender.Commit()
+	return checkErr
+}
+
+func (c *Check) doRun(staticTags []string) ([]string, error) {
+	tags := copyStrings(staticTags)
+
+	// Create connection
+	connErr := c.session.Connect()
+	if connErr != nil {
+		return tags, fmt.Errorf("snmp connection error: %s", connErr)
+	}
+	defer func() {
+		err := c.session.Close()
+		if err != nil {
+			log.Warnf("failed to close session: %v", err)
+		}
+	}()
+
+	// If no OIDs, try to detect profile using device sysobjectid
+	if !c.config.oidConfig.hasOids() {
+		sysObjectID, err := fetchSysObjectID(c.session)
+		if err != nil {
+			return tags, fmt.Errorf("failed to fetching sysobjectid: %s", err)
+		}
+		profile, err := getProfileForSysObjectID(c.config.profiles, sysObjectID)
+		if err != nil {
+			return tags, fmt.Errorf("failed to get profile sys object id for `%s`: %s", sysObjectID, err)
+		}
+		err = c.config.refreshWithProfile(profile)
+		if err != nil {
+			// Should not happen since the profile is one of those we matched in getProfileForSysObjectID
+			return tags, fmt.Errorf("failed to refresh with profile `%s` detected using sysObjectID `%s`: %s", profile, sysObjectID, err)
+		}
+	}
+	tags = append(tags, c.config.profileTags...)
+
+	// Fetch and report metrics
+	if c.config.oidConfig.hasOids() {
+		c.config.addUptimeMetric()
```

Comment:
That's a good point. It's possible to avoid that as you suggested, but that might add more complexity elsewhere.

For example, oidConfig.hasOids won't work anymore and we might need to add something like `-1` to account for `addUptimeMetric`.

```
func (oc *oidConfig) hasOids() bool {
	return len(oc.columnOids) > 0 || len(oc.scalarOids) > 1
}
```

For now, the logic is same as what's done in Python, I did it that way to make it easier to have the same behaviour, but this part could be improved (I think could be done later).